fix: only show editable orgs on deployment page

code-asher opened the pull request:
Also make sure the redirect from /organizations goes to an org that the user can edit, rather than always the default org.

code-asher committed:
fix: only show editable orgs on deployment page

Also make sure the redirect from /organizations goes to an org that the
user can edit, rather than always the default org.

code-asher committed:
canEditOrganization was lying

Emyrk commented:
If you cannot `update` the organization, should we show `Organization Summary`?

Emyrk commented on site/src/api/queries/organizations.ts:
```diff
@@ -121,6 +124,60 @@ export const provisionerDaemons = (organization: string) => {
+    },
+    action: "update",
+  },
+  auditOrganization: {
+    object: {
+      resource_type: "audit_log",
+      organization_id: organizationId,
+    },
+    action: "read",
+  },
```
If we move auditing back to it's own page outside this settings page, we can drop this one :+1:

code-asher replied:
Will do!!

code-asher commented:
> If you cannot update the organization, should we show Organization Summary?

Yup I want to do this, I was going to do it in a separate PR though!

Emyrk commented:
> Yup I want to do this, I was going to do it in a separate PR though! For now it shows an ineditable form, as before.

Perfect :+1:. Mainly just curious

code-asher committed:
Limit the number of org permission checks

We show the org on the sidebar if they can edit anything, and we show
each sub-link if they can view it, which means we were making both edit
and view permission checks.

Instead, show each link if they can edit it (not just view), which
negates the need for separate view permissions.

Incidentally, this also reduces the number of checks we need to make for
individual pages, since some of them were only used on the sidebar.

code-asher commented:
The new summary page: https://github.com/coder/coder/pull/14214

Kinda sad and sparse looking but hopefully better than an ineditable form :stuck_out_tongue:

code-asher committed:
fixup! fix: only show editable orgs on deployment page

aslilac commented on site/src/api/queries/organizations.ts:
```diff
@@ -156,28 +158,91 @@ export const organizationPermissions = (organizationId: string | undefined) => {
+ * If organizations are undefined, return a disabled query.
+ */
+export const organizationsPermissions = (
+  organizations: Organization[] | undefined,
```
since this only needs the `id`, can we just have it take a `string[]` instead and do a `.map` when we call it?

code-asher replied:
Yup, works for me

aslilac commented on site/src/api/queries/organizations.ts:
```diff
@@ -156,28 +158,91 @@ export const organizationPermissions = (organizationId: string | undefined) => {
+          const index = key.indexOf(".");
+          const orgId = key.substring(0, index);
+          const perm = key.substring(index + 1);
+          if (!acc[orgId]) {
+            acc[orgId] = { [perm]: value };
+          } else {
+            acc[orgId][perm] = value;
+          }
```
```suggestion
          if (!acc[orgId]) {
            acc[orgId] = {};
          }
          acc[orgId][perm] = value;
```

code-asher replied:
Truuuuueeeee this is probably better lol, I had it like that and changed it but for the life of me I cannot justify it.

I also considered:

```
acc[orgId] ??= {}
acc[orgId][perm] = value;
```

But I think maybe it looks too magical

aslilac commented on site/src/pages/ManagementSettingsPage/OrganizationSettingsPage.tsx:
```diff
@@ -32,37 +35,43 @@ const OrganizationSettingsPage: FC = () => {
-      return <Navigate to={`/organizations/${defaultOrg.name}`} replace />;
+    const editableOrg = organizations
+      .sort((a, b) => {
+        // Prefer default org (it may not be first).
+        if (a.is_default) {
+          return -1;
+        } else if (b.is_default) {
+          return 1;
+        }
+        return 0;
```
little known, and kind of cursed, but...

```
> true - true
0
> true - false
1
> false - true
-1
```

so you could just do...

```suggestion
        // Prefer default org (it may not be first).
        return b.is_default - a.is_default
```

which is also technically more correct, because right now you're not properly checking the case where `a.is_default === true && b.is_default === true` (even though that should be impossible)

code-asher replied:
Ahahahaha addition/substraction on booleans feels very on-brand for JavaScript lol

Good point on them both being true, will fix

code-asher replied:
Uhh how bad is it if I have to disable a lint check to make this happen

code-asher replied:
Turns out I am dumb and can just use some ternaries

code-asher replied:
Thank you by the way, the way I did this was quite careless

aslilac commented on site/src/pages/ManagementSettingsPage/SidebarView.tsx:
```diff
@@ -13,18 +13,12 @@ import { type ClassName, useClassName } from "hooks/useClassName";
+  /** The active org name, if any.  Overrides activeSettings. */
+  activeOrganizationName: string | undefined;
+  /** Organizations and their permissions or undefined if still fetching. */
+  organizations: [Organization, AuthorizationResponse][] | undefined;
```
not a fan of tuples like this. could we do something like...

```ts
interface OrganizationWithPermissions extends Organization {
  permissions: AuthorizationResponse;
}
```

...instead?

code-asher replied:
Yup, works for me!  Yeah the tuple thing feels kinda weird in JavaScript

aslilac commented on site/src/pages/ManagementSettingsPage/SidebarView.tsx:
```diff
@@ -167,19 +133,77 @@ function urlForSubpage(organizationName: string, subpage: string = ""): string {
+ */
+const OrganizationsSettingsNavigation: FC<
+  OrganizationsSettingsNavigationProps
+> = (props) => {
+  // Wait for organizations and their permissions to load in.
+  if (!props.organizations) {
+    return <Loader />;
+  }
+
+  if (
+    props.organizations.length <= 0 &&
+    !props.permissions.createOrganization
+  ) {
+    return null;
+  }
```
also we typically destructure `props` in our React components rather than accessing it directly, and I like the consistency. this should look more like...

```suggestion
> = ({ organizations, permissions, whateverElse }) => {
  // Wait for organizations and their permissions to load in.
  if (!organizations) {
    return <Loader />;
  }

  if (
    organizations.length <= 0 &&
    !permissions.createOrganization
  ) {
    return null;
  }
```

code-asher replied:
Awww bummer, I am the opposite.  Will change for consistency though.

aslilac commented on site/src/pages/ManagementSettingsPage/SidebarView.tsx:
```diff
@@ -200,8 +224,7 @@ const OrganizationSettingsNavigation: FC<
       </SidebarNavItem>
-      {props.active && !props.permissions && <Loader />}
-      {props.active && props.permissions && (
+      {props.active && (
```
same in this component

code-asher committed:
Pass in ids instead of entire orgs

code-asher committed:
Assign permission value in one spot

code-asher committed:
Improved sort

code-asher committed:
Tuples-b-gone

code-asher committed:
Destruct props

aslilac commented on site/src/api/queries/organizations.ts:
```diff
@@ -156,28 +157,90 @@ export const organizationPermissions = (organizationId: string | undefined) => {
+  }
+
+  return {
+    queryKey: ["organizations", "permissions"],
```
`queryKey` needs to include the `organizationIds` in it

code-asher replied:
Ahh for some reason I thought this would get invalidated automatically if the organizations query changed, but I think actually an invalidation has to be added manually?  Would that work too or do you think adding the IDs is better?

code-asher replied:
Actually idk what i am thinking, easy to just throw the ids on there :stuck_out_tongue:

aslilac commented on site/src/pages/ManagementSettingsPage/OrganizationSettingsPage.test.tsx:
```diff
@@ -0,0 +1,125 @@
+const renderRootPage = async () => {
+  renderWithManagementSettingsLayout(<OrganizationSettingsPage />, {
+    route: "/organizations",
+    path: "/organizations",
```
pretty sure you can just change this to...
```suggestion
    path: "/organizations/:organization?",
```

...and then get rid of the `extraRoutes`

code-asher replied:
Oh clever!

aslilac commented on site/src/pages/ManagementSettingsPage/OrganizationSettingsPage.tsx:
```diff
@@ -32,37 +35,42 @@ const OrganizationSettingsPage: FC = () => {
+        // Prefer default org (it may not be first).
+        // JavaScript will happily subtract booleans, but use numbers to keep
+        // the compiler happy.
+        return (b.is_default ? 1 : 0) - (a.is_default ? 1 : 0);
```
oh, this is great!

aslilac commented on site/src/pages/ManagementSettingsPage/Sidebar.tsx:
```diff
@@ -14,27 +17,45 @@ import { SidebarView } from "./SidebarView";
+        permissions: orgPermissionsQuery.data?.[org.id],
+      };
+    })
+    // TypeScript is not able to infer whether permissions are defined from the
+    // canEditOrganization call, so although redundant this helps figure it out.
+    .filter(
+      (org): org is OrganizationWithPermissions =>
+        org.permissions !== undefined,
+    )
+    .filter((org) => {
+      return canEditOrganization(org.permissions);
+    });
```
could you just do...

```suggestion
    .filter(
      (org) => org.permissions && canEditOrganization(org.permissions);
    });
```

...instead? or is it still not happy with that? I try to avoid consecutive `.filter` calls

code-asher replied:
As fortune would have it, I did indeed try exactly this first but sadly no dice.  It really seems like it should have worked though.

Tried it again now just in case I had messed it up the first time, but sadly:

```
Type '{ permissions: AuthorizationResponse | undefined; description: string; created_at: string; updated_at: string; is_default: boolean; id: string; name: string; display_name: string; icon: string; }[] | undefined' is not assignable to type 'OrganizationWithPermissions[] | undefined'.
```

code-asher replied:
Oh wait duh I can just move `is OrganizationWithPermissions` to the second filter and get rid of the first.

code-asher committed:
No need for extra routes with a bit of ?

code-asher committed:
Add org IDs to permissions query

code-asher committed:
Two is now one

code-asher committed:
Merge remote-tracking branch 'github/main' into asher/show-editable-orgs

code-asher committed:
fixup! Merge remote-tracking branch 'github/main' into asher/show-editable-orgs
